File: gen_commits.go
package main

import (
	"bufio"
	"crypto/sha1"
	"encoding/binary"
	"encoding/hex"
	"fmt"
	"os"
	"strconv"
	"strings"
)

var tree = "tree 6a0165c2aea6cfc5fba01029ede7a8da6c85f6f6"
var author = "author Ben Thayer <ben@benthayer.com> 1599715620 -0500"
var committer = "committer Ben Thayer <ben@benthayer.com> 1599715620 -0500"
var message = "Initial commit"

// 1<<28 == 16^7
const num_total_hashes = 1 << 28

func gen_hash(parent_hash string, salt_int int) [20]byte {
	parent := "parent " + parent_hash
	salt := strconv.Itoa(salt_int)
	commit := strings.Join(
		[]string{
			tree,
			parent,
			author,
			committer,
			"",
			message,
			"",
			salt,
			""},
		"\n")

	prefix := "commit " + strconv.Itoa(len(commit)) + "\x00"
	object := prefix + commit
	data := []byte(object)
	sha_sum := sha1.Sum(data)
	return sha_sum
}

func check(e error) {
	if e != nil {
		panic(e)
	}
}

func sum_to_int(sha_sum [20]byte) uint32 {
	hash := make([]byte, 4)
	copy(hash, sha_sum[:4])
	hash[3] = hash[3] - (hash[3] % 16)
	return binary.BigEndian.Uint32(hash) / 16
}

type CommitInfo struct {
	sha_sum [20]byte
	salt    int
}

func saltMine(parent string, commit_generated *[num_total_hashes]bool, salt_chan chan int, result_chan chan CommitInfo) {
	for salt := range salt_chan {
		sha_sum := gen_hash(parent, salt)

		// Try again if it's unique
		if commit_generated[sum_to_int(sha_sum)] {
			continue
		}

		// Tell the others or die trying!
		select {
		case result_chan <- CommitInfo{sha_sum, salt}:
			return
		}
	}
}

func get_next_commit(parent string, commit_generated *[num_total_hashes]bool, commit_number int) CommitInfo {

	var numWorkers int
	if commit_number <= num_total_hashes/2 {
		numWorkers = 1
	} else if commit_number <= num_total_hashes/4 {
		numWorkers = 2
	} else if commit_number <= num_total_hashes/8 {
		numWorkers = 4
	} else {
		numWorkers = 8
	}

	salt := 1
	salt_chan := make(chan int)
	result_chan := make(chan CommitInfo, numWorkers)

	for i := 0; i < numWorkers; i++ {
		go saltMine(parent, commit_generated, salt_chan, result_chan)
	}

	for {
		select {
		case salt_chan <- salt:
			salt++
		case result := <-result_chan:
			commit_generated[sum_to_int(result.sha_sum)] = true
			close(salt_chan)
			return result
		}
	}
}

func main() {
	// Track the commits we're making
	f, err := os.Create("./commits")
	check(err)
	defer f.Close()
	w := bufio.NewWriter(f)

	// progress tracking
	var expected float64 = 0
	var progress float64 = 0

	for commit_number := 0; commit_number < (num_total_hashes); commit_number++ {
		expected += float64(num_total_hashes) / float64(num_total_hashes-commit_number)
	}

	// variables
	var commit_generated [num_total_hashes]bool

	// set up first commit
	parent_hash := "f9a17849fe28dff34647f698a392be2a9ce3617b"
	hash_decoded, err := hex.DecodeString(parent_hash[:])
	check(err)
	var sha_sum [20]byte
	copy(sha_sum[:], hash_decoded)
	commit_generated[sum_to_int(sha_sum)] = true

	// start generating commits
	for commit_number := 1; commit_number < (num_total_hashes); commit_number++ {
		commit_info := get_next_commit(parent_hash, &commit_generated, commit_number)

		parent_hash = hex.EncodeToString(commit_info.sha_sum[:])

		// progress tracking
		w.WriteString(parent_hash + " " + strconv.Itoa(commit_info.salt) + "\n")
		progress += float64(num_total_hashes) / float64(num_total_hashes-commit_number)
		if commit_number > (num_total_hashes - 12000) {
			w.Flush()
		}
		fmt.Print(strconv.FormatFloat(progress/expected, 'f', 8, 64) + "   " + strconv.Itoa(commit_number) + "\r")
	}
}

File: rust/src/gen_commits.rs
/// Since use of `select!` macros are unstable, we MUST switch to using futures
/// instead.
// use crossbeam_channel::{bounded, select, unbounded, Receiver, Sender};
// ---------------------------
use tokio; // go through tutorial -> https://tokio.rs/tokio/tutorial

// use futures::{executor::block_on, future, select as fut_select};
use threadpool::ThreadPool;

use byteorder::{BigEndian, ByteOrder};
use ring::digest;
use std::convert::TryFrom;

// 1<<28 == 16^7
pub const NUM_TOTAL_HASHES_usize: usize = 1 << 28;
pub const NUM_TOTAL_HASHES_i32: i32 = 1 << 28;

pub fn gen_hash(parent_hash: &str, salt_int: i32) -> Vec<u8> {
  let tree = "tree 6a0165c2aea6cfc5fba01029ede7a8da6c85f6f6";
  let author = "author Ben Thayer <ben@benthayer.com> 1599715620 -0500";
  let committer = "committer Ben Thayer <ben@benthayer.com> 1599715620 -0500";
  let message = "Initial commit";

  let parent = "parent ".to_string() + parent_hash;
  let salt = salt_int.to_string();
  let commit = [tree, &parent, author, committer, "", message, "", &salt, ""].join("\n");

  let prefix = "commit ".to_string() + &commit.len().to_string() + "\x00";
  let object = prefix + &commit;
  let _data = object.as_bytes();
  let sha_sum = digest::digest(&digest::SHA1_FOR_LEGACY_USE_ONLY, _data);

  return sha_sum.as_ref().to_vec();
}

type Err<T> = std::result::Result<T, Error>;

struct Error;

/// For unrecoverable errors -- perhaps need custom error types for each unique
/// contextual error with this program. See the link below for more details:
/// http://web.mit.edu/rust-lang_v1.25/arch/amd64_ubuntu1404/share/doc/rust/html/book/first-edition/error-handling.html
// pub fn check(option: Option<T>) {
//   if error != None {
//     panic!(error);
//   }
// }

pub fn sum_to_int(sha_sum: Vec<u8>) -> u32 {
  let mut dst = vec![0; 4];

  // for reference, see the link below to understand why this works:
  // https://stackoverflow.com/questions/28219231/how-to-idiomatically-copy-a-slice
  dst.clone_from_slice(&sha_sum[0..4]);
  dst[3] = dst[3] - (dst[3] % 16);

  let int_to_return: u32 = BigEndian::read_u32(&dst) / 16;

  return int_to_return;
}

pub struct CommitInfo {
  pub sha_sum: Vec<u8>,
  pub salt: i32,
}

pub fn salt_mine(
  parent: &str,
  commit_generated: [bool; NUM_TOTAL_HASHES_usize],
  salt_work_receiver: Receiver<i32>,
  result_sender: Sender<CommitInfo>,
) {
  for salt in salt_work_receiver {
    let sha_sum = gen_hash(parent, salt);
    let index: usize = usize::try_from(sum_to_int(sha_sum)).unwrap();

    if commit_generated[index] {
      continue;
    }

    // Tell the others, or die trying!
    // `select!` macro is unstable.
    // (see reference: https://stackoverflow.com/questions/52302371/what-is-the-equivalent-of-gos-select-case-paradigm-for-channels-in-rust)

    // select! {
    //   send(result_sender, CommitInfo { sha_sum, salt });
    //   // recv(result_receiver, res) -> res => return res,
    // }
  }
}

pub fn get_next_commit(
  parent: &str,
  commit_generated: [bool; NUM_TOTAL_HASHES_usize],
  commit_number: i32,
) -> CommitInfo {
  let numWorkers: i32;

  if commit_number <= (NUM_TOTAL_HASHES_i32) / 2 {
    numWorkers = 1;
  } else if commit_number <= (NUM_TOTAL_HASHES_i32) / 4 {
    numWorkers = 2;
  } else if commit_number <= (NUM_TOTAL_HASHES_i32) / 8 {
    numWorkers = 4;
  } else {
    numWorkers = 8;
  }

  // Create a new thread pool capable of executing four jobs concurrently.
  let pool = ThreadPool::new(4);

  let salt: i32 = 1;
  /// Use `futures` instead.
  // let (salt_work_sender, salt_work_receiver) = unbounded();
  // let (result_sender, result_receiver) = bounded(numWorkers as usize);
  let fut_salt_values = async {
    let fut_salt_mine_result = async move {
      for i in 0..numWorkers {
        let result_sender = result_sender.clone();

        pool.execute(move || {
          salt_mine(parent, commit_generated, salt_work_receiver, result_sender);
        });
      }
    };

    // Wait until salt has been mined before mining another.
    fut_salt_mine_result.await;
  };

  // Use of `select!` macro is unstable.

  // Select must operate concurrently with `salt_mine` so that `result_receiver`
  // is concurrently receiving input from the sending-channel output of
  // `saltMine()`.
  // select! {
  //   /// Learn how to properly use the `select!` macro in channels
  //   salt_work_sender.send(salt),
  //   salt += 1,
  //   block_on(fut_salt_values), // runs asynchronously, waiting for new salt

  //   commit_generated[sum_to_int(result_receiver.sha_sum)] = true,
  //   close(salt_chan),
  //   recv(result_receiver, result) => return result,
  // }
}
